cases/src/org/commcare/cases/instance/CaseDataInstance.java

```diff
@@ -0,0 +1,84 @@
+package org.commcare.cases.instance;
+
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.model.instance.utils.FormLoadingUtils;
+import org.javarosa.core.services.Logger;
+import org.javarosa.xml.util.InvalidStructureException;
+
+import java.io.IOException;
+
+/**
+ * An external data instance that respects CaseDB template specifications.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class CaseDataInstance extends ExternalDataInstance {
+    private static TreeElement caseDbSpecTemplate;
+    private static final String CASEDB_WILD_CARD = "CASEDB_WILD_CARD";
+
+    static {
+        try {
+            caseDbSpecTemplate =
+                    FormLoadingUtils.xmlToTreeElement("/casedb_instance_structure.xml");
+        } catch (InvalidStructureException e) {
+            caseDbSpecTemplate = null;
+        } catch (IOException e) {
+            caseDbSpecTemplate = null;
+        }
+    }
+
+    public CaseDataInstance() {
+        // For serialization
+    }
+
+    private CaseDataInstance(String reference, String instanceid) {
+        super(reference, instanceid);
+    }
+
+    public CaseDataInstance buildExternalDataInstance(String reference, String instanceId) {
+        return new CaseDataInstance(reference, instanceId);
+    }
+
+    /**
+     * Does the reference follow the statically defined CaseDB spec?
+     */
+    public boolean hasTemplatePath(TreeReference ref) {
+        if (caseDbSpecTemplate != null) {
+            return followsTemplateSpec(ref, caseDbSpecTemplate, 0);
+        } else {
+            // failed to load casedb spec template, default to super implementation
+            Logger.log("CaseDb Warning",
+                    "Using default hasTemplatePath implementation: reference resolution will break!");
+            return super.hasTemplatePath(ref);
```

Comment:
We will end up touching this class (loading into permgen space) when/if we do any reflected walks through the code (like we do in CCODK when we build the prototype factory.

Is the inherent advantage here (v. lazy init) that it occurs in a threadsafe way? I think in general it's ~always safer to lazy init this kind of thing whenever possible, since the semantics are much clearer and if there's a failure it happens explicitly when you requested an operation that used the class.